packages/commonwealth/client/scripts/views/pages/general_contract/index.tsx

```diff
@@ -0,0 +1,394 @@
+              <CWText>State Mutability</CWText>
+              <CWText>Inputs</CWText>
+              <CWText>Outputs</CWText>
+              <CWText>Call Function</CWText>
+            </div>
+            {loadContractAbi().map((fn: AbiItem, fnIdx: number) => {
+              return (
+                <div class="function-row">
+                  <CWText>{fn.name}</CWText>
+                  <CWText>{fn.stateMutability}</CWText>
+                  <div class="functions-input-container">
+                    {fn.inputs.map((input: AbiInput, inputIdx: number) => {
+                      return (
+                        <div>
+                          <div class="function-inputs">
+                            <CWText>[{inputIdx}]</CWText>
+                            <CWText>{input.type}</CWText>
+                            <CWText>{input.name}</CWText>
+                          </div>
+                          <div>
+                            <CWTextInput
+                              name="Contract Input Field"
+                              placeholder="Insert Input Here"
+                              oninput={(e) => {
+                                if (
+                                  !this.state.form.functionNameToFunctionInputArgs.has(
+                                    fn.name
+                                  )
+                                ) {
+                                  this.state.form.functionNameToFunctionInputArgs.set(
+                                    fn.name,
+                                    new Map<number, string>()
+                                  );
+                                  console.log('setting new map');
+                                  const inputArgMap =
+                                    this.state.form.functionNameToFunctionInputArgs.get(
+                                      fn.name
+                                    );
+                                  inputArgMap.set(inputIdx, e.target.value);
+                                  this.state.form.functionNameToFunctionInputArgs.set(
+                                    fn.name,
+                                    inputArgMap
+                                  );
+                                } else {
+                                  const inputArgMap =
+                                    this.state.form.functionNameToFunctionInputArgs.get(
+                                      fn.name
+                                    );
+                                  inputArgMap.set(inputIdx, e.target.value);
+                                  this.state.form.functionNameToFunctionInputArgs.set(
+                                    fn.name,
+                                    inputArgMap
+                                  );
+                                }
+                                this.state.loaded = true;
+                              }}
+                              inputValidationFn={(
```

Comment:
all of this validation stuff should be pulled into its own function and just called here. it's too much in the UI